### PR TITLE
fix!: OpDef serialisation

### PR DESCRIPTION
**Pull request by doug-q**
BREAKING CHANGE:
* OpRef serialisation schema

BEGIN_COMMIT_OVERRIDE
fix: `OpDef` serialization (#1013)

BREAKING CHANGE: New serialisation schema
END_COMMIT_OVERRIDE

### CODE REVIEW

File: hugr-py/src/hugr/serialization/ops.py [hunk 1]
```diff
@@ -218,6 +209,11 @@ def insert_port_types(self, in_types: TypeRow, out_types: TypeRow) -> None:
         assert len(in_types) == 0
         self.types = list(out_types)
 
+    # class Config:
+    #     json_schema_extra = {
+    #         "required": ["op", "parent", "types"]
+    #     }
+
```
Comment (mark-koch) [inline hugr-py/src/hugr/serialization/ops.py:216]:
```suggestion
```

File: hugr-py/src/hugr/serialization/ops.py [hunk 2]
```diff
@@ -501,24 +497,29 @@ class OpType(RootModel):
...
     description: str  # Human readable description of the operation.
-    inputs: list[tuple[str | None, Type]]
-    outputs: list[tuple[str | None, Type]]
-    misc: dict[str, Any]  # Miscellaneous data associated with the operation.
-    def_: str | None = Field(
-        ..., alias="def"
-    )  # (YAML?)-encoded definition of the operation.
-    extension_reqs: ExtensionSet  # Resources required to execute this operation.
+    misc: Optional[dict[str, Any]] = None
+    signature: Optional[PolyFuncType] = None
```
Comment (mark-koch) [inline hugr-py/src/hugr/serialization/ops.py:521]:
```suggestion
    misc: dict[str, Any] | None = None
    signature: PolyFuncType | None = None
```

**Review by mark-koch — APPROVED** [review]

File: hugr-py/src/hugr/serialization/ops.py (hunk 2 above)
Comment (mark-koch) [reply hugr-py/src/hugr/serialization/ops.py]: The remaining code usues `|` so it's more consistent. Also https://docs.astral.sh/ruff/rules/non-pep604-annotation/ so I'm confused that Ruff didn't reformat it itself

File: hugr/src/extension/op_def.rs [hunk 3]
```diff
@@ -495,7 +502,10 @@ mod test {
         let type_scheme = PolyFuncType::new(vec![TP], FunctionType::new_endo(vec![list_of_var]));
 
         let def = e.add_op(OP_NAME, "desc".into(), type_scheme)?;
-        def.add_lower_func(LowerFunc::FixedHugr(ExtensionSet::new(), Hugr::default()));
+        def.add_lower_func(LowerFunc::FixedHugr {
+            extensions: ExtensionSet::new(),
+            hugr: Hugr::default(),
```
Comment (acl-cqc) [inline hugr/src/extension/op_def.rs:507]: If this is the test I should look at - no this doesn't look good! This is a Hugr that we can splice in in place of an Op (Node), so should be a DFG-rooted thing with the same inputs/outputs as specified in the OpDef's type_scheme....

File: hugr/src/extension/op_def.rs [hunk 4]
```diff
@@ -265,11 +265,19 @@ impl Debug for SignatureFunc {
...
 /// that implements the operation using a set of other extensions.
 #[derive(serde::Deserialize, serde::Serialize)]
+#[serde(untagged)]
 pub enum LowerFunc {
     /// Lowering to a fixed Hugr. Since this cannot depend upon the [TypeArg]s,
     /// this will generally only be applicable if the [OpDef] has no [TypeParam]s.
-    #[serde(rename = "hugr")]
-    FixedHugr(ExtensionSet, Hugr),
+    FixedHugr {
+        /// The extensions required by the [`Hugr`]
```
Comment (acl-cqc) [inline hugr/src/extension/op_def.rs:273]: Hmmmm, perhaps to think about in the future, but if we insist that the `Hugr` is DFG-rooted, then we can extract the ExtensionSet from the root node's `extension_reqs`, so I'm not sure we really need this...

**Review by acl-cqc — APPROVED** [review]
Thanks!

**Pull request merged**